internal/invocation/invocation_service.go

```diff
@@ -17,68 +17,147 @@
 package invocation
 
 import (
+	"context"
 	"fmt"
+	"sync/atomic"
+	"time"
 
+	"github.com/hazelcast/hazelcast-go-client/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/internal/cb"
 	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
 	"github.com/hazelcast/hazelcast-go-client/internal/proto"
 )
 
+const (
+	ready   = 0
+	stopped = 1
+)
+
 type Handler interface {
 	Invoke(invocation Invocation) (groupID int64, err error)
 }
 
 type Service struct {
-	requestCh       <-chan Invocation
-	urgentRequestCh <-chan Invocation
-	responseCh      <-chan *proto.ClientMessage
+	requestCh       chan Invocation
+	urgentRequestCh chan Invocation
+	responseCh      chan *proto.ClientMessage
 	// removeCh carries correlationIDs to be removed
-	removeCh    <-chan int64
+	removeCh    chan int64
 	doneCh      chan struct{}
 	invocations map[int64]Invocation
 	handler     Handler
 	logger      ilogger.Logger
+	state       int32
 }
 
 func NewService(
-	requestCh <-chan Invocation,
-	urgentRequestCh <-chan Invocation,
-	responseCh <-chan *proto.ClientMessage,
-	removeCh <-chan int64,
 	handler Handler,
 	logger ilogger.Logger) *Service {
 	service := &Service{
-		requestCh:       requestCh,
-		urgentRequestCh: urgentRequestCh,
-		responseCh:      responseCh,
-		removeCh:        removeCh,
+		requestCh:       make(chan Invocation),
+		urgentRequestCh: make(chan Invocation),
+		responseCh:      make(chan *proto.ClientMessage),
+		removeCh:        make(chan int64),
 		doneCh:          make(chan struct{}),
 		invocations:     map[int64]Invocation{},
 		handler:         handler,
 		logger:          logger,
+		state:           ready,
 	}
 	go service.processIncoming()
 	return service
 }
 
 func (s *Service) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.state, ready, stopped) {
+		return
+	}
 	close(s.doneCh)
+	close(s.requestCh)
+	close(s.urgentRequestCh)
+	close(s.responseCh)
+	close(s.removeCh)
 }
 
 func (s *Service) SetHandler(handler Handler) {
 	s.handler = handler
 }
 
+func (s *Service) SendRequest(ctx context.Context, inv Invocation) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("sending invocation: %w", ctx.Err())
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("sending urgent invocation: %w", hzerrors.ErrClientNotActive))
+	case s.requestCh <- inv:
```

Comment:
There is race between `SendRequest` and `Stop` in which we can end up with writing to a closed channel. 
Lets say that we have two goroutines A and B
- From A, I called `SendRequest`
- We have checked that state is ready and we are about to run `select` statement. 
- From B, I called Stop. Change state to `stopped`
- And continued and closed all the channels. 
- A continues on select statement which will randomly select a case. If it selects the `s.requestCh <- inv:` , `PANIC`

Reply:
I was only able to produce this by adding a sleep right before `select` in `SendRequest` so there's no test for this. But I've removed the lines which closes channels (except `doneCh`) since closing them is not required in this case: https://github.com/hazelcast/hazelcast-go-client/pull/655/commits/08fb8d091789a14d402521047a9e7fb88dd997fa

Reply:
We still have problems. It looks like we can end up with hanging invocations. 
Again lets say that we have three goroutines `A` , `B` and `C`
- `C` is the goroutine that runs `processIncoming` 
- From `A` I called `SendRequest`
- We have checked that state is ready and I am waiting on `requestCh` for `C` to take the request.
- From `B` I called `Stop`. Changed the state and close the `doneCh`
- `C` runs the `select` statement which will pick a random one. If it selects the `doneCh` , we will break the loop. 
and it seems like no one will consume from `requestCh` anymore. The problem seems to be same all other channels read via `processIncoming` . A sender to these channels can wait indefinitely. 
Make sense ?

Reply:
No, the only way of writing to request channel is calling `SendRequst`, and it has the following `select`:
```
select {
case <-ctx.Done():
	return fmt.Errorf("sending invocation: %w", ctx.Err())
case <-s.doneCh:
	return cb.WrapNonRetryableError(fmt.Errorf("sending urgent invocation: %w", hzerrors.ErrClientNotActive))
case s.requestCh <- inv:
	return nil
}
```
Since `doneCh` is closed, `SendRequest` will return an error.

Reply:
Ok my mistake. I fail to reason about the blocking calls on the select statements. It is new to me.  Since `requestCh` is not buffered, it will be unblocked via `case <-s.doneCh:` when channel is closed.